bolt-modules/boltlib/lib/puppet/datatypes/target.rb

```diff
@@ -1,22 +1,55 @@
 # frozen_string_literal: true
 
 Puppet::DataTypes.create_type('Target') do
-  interface <<-PUPPET
-    attributes => {
-      uri => String[1],
-      options => { type => Hash[String[1], Data], value => {} }
-    },
-    functions => {
-      name => Callable[[], String[1]],
-      host => Callable[[], Optional[String]],
-      password => Callable[[], Optional[String[1]]],
-      port => Callable[[], Optional[Integer]],
-      protocol => Callable[[], Optional[String[1]]],
-      user => Callable[[], Optional[String[1]]],
-    }
-  PUPPET
-
+  begin
+    inventory = Puppet.lookup(:bolt_inventory)
+    inventory_version = inventory.version
+    if inventory_version != 1
+      target_implementation_class = inventory.target_implementation_class
+    end
+  rescue Puppet::Context::UndefinedBindingError
+    inventory_version = 1
+  end
   load_file('bolt/target')
 
-  implementation_class Bolt::Target
+  if inventory_version == 1
+    interface <<-PUPPET
+      attributes => {
+        uri => String[1],
+        options => { type => Hash[String[1], Data], value => {} }
+      },
+      functions => {
+        name => Callable[[], String[1]],
+        host => Callable[[], Optional[String]],
+        password => Callable[[], Optional[String[1]]],
+        port => Callable[[], Optional[Integer]],
+        protocol => Callable[[], Optional[String[1]]],
+        user => Callable[[], Optional[String[1]]],
+      }
+    PUPPET
+    implementation_class Bolt::Target
+  else
+    interface <<-PUPPET
+      attributes => {
+        uri => { type => Optional[String[1]], kind => given_or_derived },
+        name => { type => Optional[String[1]] , kind => given_or_derived },
+        target_alias => { type => Optional[Variant[String[1], Array[String[1]]]], kind => given_or_derived },
```

Comment:
I can't find / don't remember the discussion where we decided to call this `target_alias`? I assumed it was `alias`, and the error message is readable but not pretty:
```
lucy ~/githubs/bolt (GH-1125) $ be bolt plan run test --params @data.json
Starting: plan test
Finished: plan test in 0.01 sec
{
  "kind": "bolt/pal-error",
  "msg": "'new_Target' expects one of:\n  (Struct[{'uri' => Optional[String[1]], 'name' => Optional[String[1]], 'target_alias' => Optional[Variant[String[1], Array[String[1]]]], 'config' => Optional[Hash[String[1], Data]], 'vars' => Optional[Hash[String[1], Data]], 'facts' => Optional[Hash[String[1], Data]], 'features' => Optional[Array[String[1]]], 'plugin_hooks' => Optional[Hash[String[1], Data]]}] hash)\n    rejected: parameter 'hash' unrecognized key 'alias'\n  (String[1] uri?, String[1] name?, Variant[String[1], Array[String[1]]] target_alias?, Hash[String[1], Data] config?, Hash[String[1], Data] facts?, Hash[String[1], Data] vars?, Array[String[1]] features?, Hash[String[1], Data] plugin_hooks?)\n    rejected: parameter 'uri' expects a value of type Undef or String, got Struct",
  "details": {
  }
}
```

Is there a reason not to just make this `alias`?

Reply:
alias is a reserved word in ruby. I dont think there is much we can do about the error message.